dask/dataframe/core.py

```diff
@@ -1130,21 +1166,33 @@ def _bind_operator_method(cls, name, op):
         def meth(self, other, level=None, fill_value=None, axis=0):
             if not level is None:
                 raise NotImplementedError('level must be None')
-            return map_partitions(op, self.column_info, self, other,
+            return map_partitions(op, self._pd, self, other,
                                   axis=axis, fill_value=fill_value)
         meth.__doc__ = op.__doc__
         bind_method(cls, name, meth)
 
     def apply(self, func, convert_dtype=True, name=no_default, args=(), **kwds):
         """ Parallel version of pandas.Series.apply """
+
         if name is no_default:
-            name = self.name
+            msg = ("name is not specified, inferred from partial data. "
+                   "Please provide name if the result is unecpected.\n"
+                   "  Before: s.apply(func)\n"
+                   "  After:  s.apply(func, name=['x', 'y']) for dataframe result\n"
+                   "  or:     s.apply(func, name='x')        for series result")
+            warnings.warn(msg)
+
+            name = _emulate(pd.Series.apply, self.head(), func,
+                            convert_dtype=convert_dtype,
+                            args=args, **kwds)
```

Comment:
What do you think about trying the function on the empty dataframe and, if it returns a good result, using the name from that result without a warning.

```python
try:
    name = func(self._pd, *args, **kwargs).name
except Exception as e:
    raise ValueError("Tried to determine name from applying the function to an empty dataframe "
                     "but received the following error:\n\t%s\n\n""
                     "Please specify the name of the resulting series as follows:\n"
                       ... % str(e))